Importing libraries 

In [33]:
import pandas as pd 
import numpy as np 


Read Json files and CSV file 

In [34]:
listings = pd.read_json('listings_final_expanded.json')
properties = pd.read_json('property_attributes_final_expanded.json')
agents = pd.read_json('agents_cleaned.json')
buyers = pd.read_json('buyers_cleaned.json')
sales=pd.read_csv('sales_cleaned.csv')

cleaning the columns 

In [35]:
def clean_code(df):
    df.columns=(df.columns.str.lower().str.strip().str.replace(' ','_'))
    return df  
 
listings=clean_code(listings)
properties=clean_code(properties)
agents=clean_code(agents)
buyers=clean_code(buyers)
sales=clean_code(sales)

In [36]:
listings.head()

listing_id         city property_type         price         sqft  \
0     L00001     New York     Apartment  1.655144e+06  2753.009121   
1     L00002  Los Angeles     Apartment  1.519141e+06  4966.988193   
2     L00003      Houston     Apartment  1.624890e+05  1267.003959   
3     L00004      Phoenix     Apartment  1.277016e+06  2128.014429   
4     L00005      Phoenix     Townhouse  5.622970e+05  4178.997421   

  date_listed agent_id   latitude   longitude  
0  2023-05-06    A0015  33.965208  -69.861589  
1  2023-02-14    A0038  42.547892  -90.277860  
2  2023-04-22    A0015  28.732327 -115.952982  
3  2024-01-02    A0042  26.403938  -74.771490  
4  2023-10-29    A0018  39.425252  -83.917878

In [37]:
properties.head()

attribute_id listing_id  bedrooms  bathrooms  floor_number  total_floors  \
0             1     L00001         5          3             9             9   
1             2     L00002         2          2            19            29   
2             3     L00003         2          3             8            26   
3             4     L00004         3          3            25            10   
4             5     L00005         2          2            15            16   

   year_built  is_rented  tenant_count furnishing_status  metro_distance_km  \
0        2001       True             4         Furnished               7.26   
1        2020      False             0       Unfurnished               4.84   
2        2007       True             4    Semi-Furnished               4.92   
3        2003      False             3         Furnished               1.20   
4        2008       True             1    Semi-Furnished               7.90   

   parking_available  power_backup  
0              False          True  
1               True         False  
2              False         False  
3              False          True  
4              False         False

In [38]:
agents.head()

agent_id         name            phone                 email  \
0    A0001  Agent A0001  +1-534-665-8373  a0001@realestate.com   
1    A0002  Agent A0002  +1-493-463-4698  a0002@realestate.com   
2    A0003  Agent A0003  +1-290-534-1121  a0003@realestate.com   
3    A0004  Agent A0004  +1-691-610-4878  a0004@realestate.com   
4    A0005  Agent A0005  +1-829-613-5411  a0005@realestate.com   

   commission_rate  deals_closed  rating  experience_years  avg_closing_days  
0             2.00            52     4.3                10                64  
1             2.20            26     3.8                20                82  
2             2.85           297     3.6                21                83  
3             1.67            31     4.5                 5                61  
4             1.11           198     4.3                16                67

In [39]:
buyers.head()

buyer_id sale_id buyer_type   payment_mode  loan_taken loan_provider  \
0         1  L01179   End User           Cash       False          None   
1         2  L00866   Investor         Cheque       False          None   
2         3  L00102   Investor         Cheque        True          Axis   
3         4  L00440   Investor  Bank Transfer       False          None   
4         5  L00059   Investor            UPI        True          HDFC   

   loan_amount  
0            0  
1            0  
2      2317757  
3            0  
4      4191221

In [40]:
sales.head()

listing_id    sale_price   date_sold  days_on_market
0     L01179  9.255800e+05  2023-07-07       65.005560
1     L00866  1.054160e+05  2023-06-14       38.004620
2     L00102  1.825184e+06  2023-09-09       22.992622
3     L00440  1.932085e+06  2023-10-29       72.012274
4     L00059  7.765860e+05  2023-05-01      116.000152

for maintain the consistent in column names we changed the caps to small, remove spaces, replaced spaces with _

There is no nested json, all are flat table structure 

Dtypes check 

In [41]:
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21200 entries, 0 to 21199
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   listing_id     21200 non-null  object 
 1   city           21200 non-null  object 
 2   property_type  21200 non-null  object 
 3   price          21200 non-null  float64
 4   sqft           21200 non-null  float64
 5   date_listed    21200 non-null  object 
 6   agent_id       21200 non-null  object 
 7   latitude       21200 non-null  float64
 8   longitude      21200 non-null  float64
dtypes: float64(4), object(5)
memory usage: 1.5+ MB


in this the date_listed is in the object type, but now we gonna convert to the date_time format

In [42]:

listings['date_listed'] = pd.to_datetime(listings['date_listed'])
print(listings['date_listed'].dtype)

datetime64[ns]


In [43]:
properties.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21200 entries, 0 to 21199
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   attribute_id       21200 non-null  int64  
 1   listing_id         21200 non-null  object 
 2   bedrooms           21200 non-null  int64  
 3   bathrooms          21200 non-null  int64  
 4   floor_number       21200 non-null  int64  
 5   total_floors       21200 non-null  int64  
 6   year_built         21200 non-null  int64  
 7   is_rented          21200 non-null  bool   
 8   tenant_count       21200 non-null  int64  
 9   furnishing_status  21200 non-null  object 
 10  metro_distance_km  21200 non-null  float64
 11  parking_available  21200 non-null  bool   
 12  power_backup       21200 non-null  bool   
dtypes: bool(3), float64(1), int64(7), object(2)
memory usage: 1.7+ MB


all columns are in the correct dtypes

In [44]:
agents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   agent_id          50 non-null     object 
 1   name              50 non-null     object 
 2   phone             50 non-null     object 
 3   email             50 non-null     object 
 4   commission_rate   50 non-null     float64
 5   deals_closed      50 non-null     int64  
 6   rating            50 non-null     float64
 7   experience_years  50 non-null     int64  
 8   avg_closing_days  50 non-null     int64  
dtypes: float64(2), int64(3), object(4)
memory usage: 3.6+ KB


all columns are in the correct dtypes

In [45]:
buyers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   buyer_id       20000 non-null  int64 
 1   sale_id        20000 non-null  object
 2   buyer_type     20000 non-null  object
 3   payment_mode   20000 non-null  object
 4   loan_taken     20000 non-null  bool  
 5   loan_provider  10034 non-null  object
 6   loan_amount    20000 non-null  int64 
dtypes: bool(1), int64(2), object(4)
memory usage: 957.2+ KB


all columns ar ein the correct dtypes, but there is some missing values in loan_provider column 

In [46]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 720 entries, 0 to 719
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   listing_id      720 non-null    object 
 1   sale_price      720 non-null    float64
 2   date_sold       720 non-null    object 
 3   days_on_market  720 non-null    float64
dtypes: float64(2), object(2)
memory usage: 22.6+ KB


here also date_sold is in string dtype, we convert into datetime format 

In [47]:

sales['date_sold'] = pd.to_datetime(sales['date_sold'])
print( sales['date_sold'].dtype)

datetime64[ns]


when we see the missing values all the columns in the dataset are no missing values, except the column "loan_provider" in buyer dataset 

In [48]:
buyers.head(15)

buyer_id sale_id buyer_type   payment_mode  loan_taken loan_provider  \
0          1  L01179   End User           Cash       False          None   
1          2  L00866   Investor         Cheque       False          None   
2          3  L00102   Investor         Cheque        True          Axis   
3          4  L00440   Investor  Bank Transfer       False          None   
4          5  L00059   Investor            UPI        True          HDFC   
5          6  L01121   End User         Cheque        True           PNB   
6          7  L00324   Investor         Cheque       False          None   
7          8  L00975   Investor         Cheque       False          None   
8          9  L00412   Investor         Cheque       False          None   
9         10  L00856   End User           Cash        True          Axis   
10        11  L00821   End User           Cash       False          None   
11        12  L00045   End User  Bank Transfer       False          None   
12        13  L00050   Investor         Cheque       False          None   
13        14  L00850   End User            UPI        True           SBI   
14        15  L00241   Investor         Cheque        True         ICICI   

    loan_amount  
0             0  
1             0  
2       2317757  
3             0  
4       4191221  
5       5583760  
6             0  
7             0  
8             0  
9       4045381  
10            0  
11            0  
12            0  
13      7930296  
14      9436023

In [49]:

loan_provider_nulls = buyers[buyers['loan_provider'].isna()]
print(f"Number of rows with null loan_provider: {len(loan_provider_nulls)}")
loan_provider_nulls

Number of rows with null loan_provider: 9966


buyer_id sale_id buyer_type   payment_mode  loan_taken loan_provider  \
0             1  L01179   End User           Cash       False          None   
1             2  L00866   Investor         Cheque       False          None   
3             4  L00440   Investor  Bank Transfer       False          None   
6             7  L00324   Investor         Cheque       False          None   
7             8  L00975   Investor         Cheque       False          None   
...         ...     ...        ...            ...         ...           ...   
19993     19994  L00830   End User            UPI       False          None   
19994     19995  L00259   Investor            UPI       False          None   
19995     19996  L00891   Investor            UPI       False          None   
19996     19997  L00834   End User            UPI       False          None   
19997     19998  L01188   Investor            UPI       False          None   

       loan_amount  
0                0  
1                0  
3                0  
6                0  
7                0  
...            ...  
19993            0  
19994            0  
19995            0  
19996            0  
19997            0  

[9966 rows x 7 columns]

Loan Providers is none to the buyers who not took the loan and their loan amount is also 0 

These None rows are also valid and should not be deleted

In [50]:

listings_cleaned = listings
properties_cleaned = properties
agents_cleaned = agents
buyers_cleaned = buyers
sales_cleaned = sales

print("All cleaned datasets declared as separate variables")

All cleaned datasets declared as separate variables


In [51]:
listings_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21200 entries, 0 to 21199
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   listing_id     21200 non-null  object        
 1   city           21200 non-null  object        
 2   property_type  21200 non-null  object        
 3   price          21200 non-null  float64       
 4   sqft           21200 non-null  float64       
 5   date_listed    21200 non-null  datetime64[ns]
 6   agent_id       21200 non-null  object        
 7   latitude       21200 non-null  float64       
 8   longitude      21200 non-null  float64       
dtypes: datetime64[ns](1), float64(4), object(4)
memory usage: 1.5+ MB


In [52]:
sales_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 720 entries, 0 to 719
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   listing_id      720 non-null    object        
 1   sale_price      720 non-null    float64       
 2   date_sold       720 non-null    datetime64[ns]
 3   days_on_market  720 non-null    float64       
dtypes: datetime64[ns](1), float64(2), object(1)
memory usage: 22.6+ KB


Connect to SQLite3

In [53]:
import sqlite3

conn = sqlite3.connect('realestate_data.db')
cursor = conn.cursor()

print("Connected to SQLite3 database: realestate_data.db")

Connected to SQLite3 database: realestate_data.db


Schema Creation

In [54]:
listings_cleaned.columns

Index(['listing_id', 'city', 'property_type', 'price', 'sqft', 'date_listed',
       'agent_id', 'latitude', 'longitude'],
      dtype='object')

In [55]:
create_listing_tables="""
CREATE TABLE IF NOT EXISTS listings (
    listing_id TEXT PRIMARY KEY,
    city TEXT,
    property_type TEXT,
    price REAL,
    sqft REAL,
    date_listed DATE,
    agent_id TEXT,
    latitude REAL,
    longitude REAL
);
"""
cursor.execute(create_listing_tables)
print("Listings table created successfully.")

Listings table created successfully.


In [56]:
properties_cleaned.columns

Index(['attribute_id', 'listing_id', 'bedrooms', 'bathrooms', 'floor_number',
       'total_floors', 'year_built', 'is_rented', 'tenant_count',
       'furnishing_status', 'metro_distance_km', 'parking_available',
       'power_backup'],
      dtype='object')

In [57]:
create_properties_table = """
CREATE TABLE IF NOT EXISTS properties (
    attribute_id INTEGER PRIMARY KEY,
    listing_id INTEGER NOT NULL,
    bedrooms INTEGER,
    bathrooms REAL,
    floor_number INTEGER,
    total_floors INTEGER,
    year_built INTEGER,
    is_rented TEXT,
    tenant_count INTEGER,
    furnishing_status TEXT,
    metro_distance_km REAL,
    parking_available TEXT,
    power_backup TEXT,
    FOREIGN KEY (listing_id) REFERENCES listings(listing_id)
)
"""

cursor.execute(create_properties_table)
print("Properties table created successfully!")

Properties table created successfully!


In [58]:
agents_cleaned.columns

Index(['agent_id', 'name', 'phone', 'email', 'commission_rate', 'deals_closed',
       'rating', 'experience_years', 'avg_closing_days'],
      dtype='object')

In [59]:
create_agents_table = """
CREATE TABLE IF NOT EXISTS agents (
    agent_id TEXT PRIMARY KEY,
    name TEXT NOT NULL,
    phone TEXT,
    email TEXT,
    commission_rate REAL,
    deals_closed INTEGER,
    rating REAL,
    experience_years INTEGER,
    avg_closing_days REAL
)
"""

cursor.execute(create_agents_table)
print("Agents table created successfully!")

Agents table created successfully!


In [60]:
buyers_cleaned.columns

Index(['buyer_id', 'sale_id', 'buyer_type', 'payment_mode', 'loan_taken',
       'loan_provider', 'loan_amount'],
      dtype='object')

In [61]:
create_buyers_table = """
CREATE TABLE IF NOT EXISTS buyers (
    buyer_id TEXT PRIMARY KEY,
    sale_id TEXT NOT NULL,
    buyer_type TEXT,
    payment_mode TEXT,
    loan_taken TEXT,
    loan_provider TEXT,
    loan_amount REAL,
    FOREIGN KEY (sale_id) REFERENCES sales(sale_id)
)
"""

cursor.execute(create_buyers_table)
print("Buyers table created successfully!")

Buyers table created successfully!


In [62]:
sales_cleaned.columns

Index(['listing_id', 'sale_price', 'date_sold', 'days_on_market'], dtype='object')

In [63]:
create_sales_table = """
CREATE TABLE IF NOT EXISTS sales (
    sale_id INTEGER PRIMARY KEY AUTOINCREMENT,
    listing_id INTEGER NOT NULL,
    sale_price REAL,
    date_sold DATE,
    days_on_market INTEGER,
    FOREIGN KEY (listing_id) REFERENCES listings(listing_id)
)
"""

cursor.execute(create_sales_table)
print("Sales table created successfully!")

Sales table created successfully!


In [64]:

listings_cleaned.to_sql('listings', conn, if_exists='append', index=False)
print(f"Inserted {len(listings_cleaned)} rows into listings table")

properties_cleaned.to_sql('properties', conn, if_exists='append', index=False)
print(f"Inserted {len(properties_cleaned)} rows into properties table")

agents_cleaned.to_sql('agents', conn, if_exists='append', index=False)
print(f"Inserted {len(agents_cleaned)} rows into agents table")

buyers_cleaned.to_sql('buyers', conn, if_exists='append', index=False)
print(f"Inserted {len(buyers_cleaned)} rows into buyers table")

sales_cleaned.to_sql('sales', conn, if_exists='replace', index=False)
print(f"Inserted {len(sales_cleaned)} rows into sales table")





Inserted 21200 rows into listings table
Inserted 21200 rows into properties table
Inserted 50 rows into agents table
Inserted 20000 rows into buyers table
Inserted 720 rows into sales table


In [65]:
print("=== DATABASE VERIFICATION ===\n")

listings_check = pd.read_sql_query("SELECT * FROM listings LIMIT 5", conn)
print("LISTINGS TABLE (Sample):")
print(listings_check)


properties_check = pd.read_sql_query("SELECT * FROM properties LIMIT 5", conn)
print("\nPROPERTIES TABLE (Sample):")
print(properties_check)


agents_check = pd.read_sql_query("SELECT * FROM agents LIMIT 5", conn)
print("\nAGENTS TABLE (Sample):")
print(agents_check)

buyers_check = pd.read_sql_query("SELECT * FROM buyers LIMIT 5", conn)
print("\nBUYERS TABLE (Sample):")
print(buyers_check)

sales_check = pd.read_sql_query("SELECT * FROM sales LIMIT 5", conn)
print("\nSALES TABLE (Sample):")
print(sales_check)


print("\n=== TOTAL ROW COUNTS ===")
for table in ['listings', 'properties', 'agents', 'buyers', 'sales']:
    count = pd.read_sql_query(f"SELECT COUNT(*) as count FROM {table}", conn)
    print(f"{table.upper()}: {count['count'][0]} rows")

conn.commit()

=== DATABASE VERIFICATION ===

LISTINGS TABLE (Sample):
  listing_id         city property_type         price         sqft  \
0     L00001     New York     Apartment  1.655144e+06  2753.009121   
1     L00002  Los Angeles     Apartment  1.519141e+06  4966.988193   
2     L00003      Houston     Apartment  1.624890e+05  1267.003959   
3     L00004      Phoenix     Apartment  1.277016e+06  2128.014429   
4     L00005      Phoenix     Townhouse  5.622970e+05  4178.997421   

           date_listed agent_id   latitude   longitude  
0  2023-05-06 00:00:00    A0015  33.965208  -69.861589  
1  2023-02-14 00:00:00    A0038  42.547892  -90.277860  
2  2023-04-22 00:00:00    A0015  28.732327 -115.952982  
3  2024-01-02 00:00:00    A0042  26.403938  -74.771490  
4  2023-10-29 00:00:00    A0018  39.425252  -83.917878  

PROPERTIES TABLE (Sample):
   attribute_id listing_id  bedrooms  bathrooms  floor_number  total_floors  \
0             1     L00001         5        3.0             9             

Data Analysis with SQL Queries

1. What is the average listing price by city?

In [66]:
query="""
SELECT city,AVG(price) AS avg_price
FROM listings GROUP BY city; """

avg_price_by_city = pd.read_sql_query(query, conn)
print(avg_price_by_city)

          city     avg_price
0      Chicago  2.430678e+06
1      Houston  2.436811e+06
2  Los Angeles  2.442418e+06
3     New York  2.493320e+06
4      Phoenix  2.459962e+06


2. What is the average price per square foot by property type?

In [67]:
query="""
SELECT property_type,AVG(price) AS avg_price
FROM listings GROUP BY property_type; """

avg_price_by_property_type = pd.read_sql_query(query, conn)
print(avg_price_by_property_type)

  property_type     avg_price
0     Apartment  2.460778e+06
1         Condo  2.427519e+06
2         House  2.470639e+06
3     Townhouse  2.452026e+06


3.  How does furnishing status impact property prices ?

In [73]:
query="""
SELECT 
    p.furnishing_status,
    AVG(l.price) AS avg_price
FROM listings l
JOIN properties p
    ON l.listing_id = p.listing_id
WHERE p.furnishing_status IS NOT NULL
GROUP BY p.furnishing_status;
"""
avg_price = pd.read_sql_query(query, conn)
print(avg_price)




  furnishing_status     avg_price
0         Furnished  2.463539e+06
1    Semi-Furnished  2.461193e+06
2       Unfurnished  2.433975e+06


4. Do properties closer to metro stations command higher prices?

In [74]:
query = """
SELECT 
    CASE 
        WHEN p.metro_distance_km <= 1 THEN 'Near Metro (<=1 km)'
        ELSE 'Far From Metro (>1 km)'
    END AS metro_category,
    AVG(l.price) AS avg_price,
    COUNT(*) AS total_properties
FROM listings l
JOIN properties p 
    ON l.listing_id = p.listing_id
WHERE p.metro_distance_km IS NOT NULL
GROUP BY metro_category;
"""

metro_price_analysis = pd.read_sql_query(query, conn)

print(metro_price_analysis)


           metro_category     avg_price  total_properties
0  Far From Metro (>1 km)  2.457369e+06             19799
1     Near Metro (<=1 km)  2.386849e+06              1401


No properties closer to metro station not higher prices in the dataset 

5. Are rented properties priced differently from non-rented ones?

In [77]:
query="""SELECT 
    p.is_rented,
    AVG(l.price) AS avg_price,
    COUNT(*) AS total_properties
FROM listings l
JOIN properties p
    ON l.listing_id = p.listing_id
WHERE p.is_rented IS NOT NULL
GROUP BY p.is_rented;
"""
rental_price_analysis = pd.read_sql_query(query, conn)
print(rental_price_analysis)


  is_rented     avg_price  total_properties
0         0  2.448250e+06             10622
1         1  2.457185e+06             10578


In [79]:
query = """
SELECT 
    p.bedrooms,
    p.bathrooms,
    AVG(l.price) AS avg_price,
    COUNT(*) AS total_properties
FROM listings l
JOIN properties p
    ON l.listing_id = p.listing_id
WHERE p.bedrooms IS NOT NULL 
  AND p.bathrooms IS NOT NULL
GROUP BY p.bedrooms, p.bathrooms
ORDER BY p.bedrooms, p.bathrooms;
"""

bed_bath_price_analysis = pd.read_sql_query(query, conn)
print(bed_bath_price_analysis)


    bedrooms  bathrooms     avg_price  total_properties
0          1        1.0  2.434001e+06              1040
1          1        2.0  2.509475e+06              1047
2          1        3.0  2.459033e+06              1028
3          1        4.0  2.455533e+06              1094
4          2        1.0  2.455735e+06              1068
5          2        2.0  2.378688e+06              1127
6          2        3.0  2.482619e+06              1060
7          2        4.0  2.555766e+06              1021
8          3        1.0  2.426008e+06              1032
9          3        2.0  2.409123e+06              1064
10         3        3.0  2.455024e+06              1030
11         3        4.0  2.478881e+06              1068
12         4        1.0  2.434956e+06              1080
13         4        2.0  2.450750e+06              1037
14         4        3.0  2.393515e+06              1034
15         4        4.0  2.471536e+06              1118
16         5        1.0  2.384940e+06           

Bedrooms and bathrooms do not significantly affect property prices in this dataset

7. Do properties with parking and power backup sell at higher prices?

In [80]:
query = """
SELECT 
    p.parking_available,
    p.power_backup,
    AVG(l.price) AS avg_price,
    COUNT(*) AS total_properties
FROM listings l
JOIN properties p
    ON l.listing_id = p.listing_id
WHERE p.parking_available IS NOT NULL
  AND p.power_backup IS NOT NULL
GROUP BY p.parking_available, p.power_backup
ORDER BY p.parking_available, p.power_backup;
"""

parking_power_analysis = pd.read_sql_query(query, conn)
print(parking_power_analysis)


  parking_available power_backup     avg_price  total_properties
0                 0            0  2.436678e+06              5408
1                 0            1  2.446293e+06              5166
2                 1            0  2.462364e+06              5383
3                 1            1  2.465652e+06              5243


Properties with parking and power backup have slightly higher average prices

8. How does year built influence listing price?

In [82]:
query = """
SELECT 
    p.year_built,
    AVG(l.price) AS avg_price
FROM listings l
JOIN properties p
    ON l.listing_id = p.listing_id
WHERE p.year_built IS NOT NULL
GROUP BY p.year_built
ORDER BY p.year_built;
"""

year_price_analysis = pd.read_sql_query(query, conn)
print(year_price_analysis)



    year_built     avg_price
0         1990  2.533792e+06
1         1991  2.573917e+06
2         1992  2.594652e+06
3         1993  2.611748e+06
4         1994  2.571505e+06
5         1995  2.434728e+06
6         1996  2.391386e+06
7         1997  2.493105e+06
8         1998  2.398159e+06
9         1999  2.498402e+06
10        2000  2.436148e+06
11        2001  2.438764e+06
12        2002  2.433615e+06
13        2003  2.395755e+06
14        2004  2.410643e+06
15        2005  2.518303e+06
16        2006  2.464741e+06
17        2007  2.379429e+06
18        2008  2.471225e+06
19        2009  2.401619e+06
20        2010  2.401120e+06
21        2011  2.440249e+06
22        2012  2.302495e+06
23        2013  2.348081e+06
24        2014  2.358477e+06
25        2015  2.457272e+06
26        2016  2.428169e+06
27        2017  2.393633e+06
28        2018  2.385865e+06
29        2019  2.561310e+06
30        2020  2.465414e+06
31        2021  2.545272e+06
32        2022  2.397387e+06
33        2023

Listing prices do not show a consistent increase with newer construction years, indicating that year built does not significantly influence pricing in this dataset.

In [85]:
query = """
SELECT city,
       AVG(price) AS median_price
FROM (
    SELECT city, price
    FROM listings
    ORDER BY city, price
)
GROUP BY city
ORDER BY median_price DESC;
"""

median_price_by_city = pd.read_sql_query(query, conn)
print(median_price_by_city)


          city  median_price
0     New York  2.493320e+06
1      Phoenix  2.459962e+06
2  Los Angeles  2.442418e+06
3      Houston  2.436811e+06
4      Chicago  2.430678e+06


New York has the highest median property price

In [86]:
query = """
SELECT 
    CASE 
        WHEN price < 2000000 THEN 'Low Price'
        WHEN price < 2500000 THEN 'Medium Price'
        ELSE 'High Price'
    END AS price_group,
    COUNT(*) AS total_properties
FROM listings
GROUP BY price_group;
"""

price_distribution = pd.read_sql_query(query, conn)
print(price_distribution)


    price_group  total_properties
0    High Price             10137
1     Low Price              9069
2  Medium Price              1994


that most properties fall into the high-price range, and very few are in the medium-price range.

What is the average days on market by city?

In [87]:
query = """
SELECT 
    l.city,
    AVG(s.days_on_market) AS avg_days_on_market
FROM listings l
JOIN sales s
    ON l.listing_id = s.listing_id
WHERE s.days_on_market IS NOT NULL
GROUP BY l.city
ORDER BY avg_days_on_market DESC;
"""

avg_days_market = pd.read_sql_query(query, conn)
print(avg_days_market)


          city  avg_days_on_market
0  Los Angeles           65.126408
1      Chicago           64.250577
2     New York           60.847639
3      Phoenix           59.654343
4      Houston           58.511086


Los Angeles has the highest average days on market,Houston has the lowest average

Which property types sell the fastest?

In [88]:
query = """
SELECT 
    l.property_type,
    AVG(s.days_on_market) AS avg_days_on_market
FROM listings l
JOIN sales s
    ON l.listing_id = s.listing_id
WHERE s.days_on_market IS NOT NULL
GROUP BY l.property_type
ORDER BY avg_days_on_market ASC;
"""

fastest_selling_types = pd.read_sql_query(query, conn)
print(fastest_selling_types)


  property_type  avg_days_on_market
0         House           58.337921
1     Apartment           60.647493
2     Townhouse           60.963765
3         Condo           66.540995


Houses sell faster, condo takes longer time to sell.

In [89]:

query = """
SELECT 
    ROUND(
        100.0 * SUM(CASE WHEN s.sale_price > l.price THEN 1 ELSE 0 END) 
        / COUNT(*), 2
    ) AS percentage_above_listing
FROM listings l
JOIN sales s
    ON l.listing_id = s.listing_id;
"""

percentage_above = pd.read_sql_query(query, conn)
print(percentage_above)


   percentage_above_listing
0                     49.31


approximately 49.31% of properties were sold above their listing price.

What is the sale-to-list price ratio by city?

In [90]:
query = """
SELECT 
    l.city,
    ROUND(AVG(s.sale_price * 1.0 / l.price), 3) AS sale_to_list_ratio
FROM listings l
JOIN sales s
    ON l.listing_id = s.listing_id
WHERE l.price IS NOT NULL 
  AND s.sale_price IS NOT NULL
GROUP BY l.city
ORDER BY sale_to_list_ratio DESC;
"""

ratio_by_city = pd.read_sql_query(query, conn)
print(ratio_by_city)


          city  sale_to_list_ratio
0      Chicago               1.002
1     New York               1.000
2  Los Angeles               1.000
3      Houston               1.000
4      Phoenix               0.999


Which listings took more than 90 days to sell?

In [92]:
query = """
SELECT *
FROM sales
WHERE days_on_market > 90;
"""

slow_listings = pd.read_sql_query(query, conn)
print(slow_listings)


    listing_id    sale_price            date_sold  days_on_market
0       L00059  7.765860e+05  2023-05-01 00:00:00      116.000152
1       L00412  1.848090e+05  2024-05-25 00:00:00      106.990553
2       L00045  4.809630e+05  2023-11-17 00:00:00       90.992445
3       L00241  8.993040e+05  2023-08-31 00:00:00       92.020419
4       L00381  3.572520e+05  2023-11-11 00:00:00      113.007285
..         ...           ...                  ...             ...
185     L00577  1.030108e+06  2024-08-07 00:00:00       94.000995
186     L00808  1.783158e+06  2024-07-28 00:00:00      100.997689
187     L00427  9.428570e+05  2024-07-21 00:00:00      100.011920
188     L00691  2.820920e+05  2024-06-18 00:00:00      101.992142
189     L00690  5.039320e+05  2023-05-13 00:00:00      118.997563

[190 rows x 4 columns]


How does metro distance affect time on market?

In [93]:
query = """
SELECT 
    p.metro_distance_km,
    AVG(s.days_on_market) AS avg_days_on_market
FROM properties p
JOIN sales s
    ON p.listing_id = s.listing_id
WHERE p.metro_distance_km IS NOT NULL
GROUP BY p.metro_distance_km
ORDER BY p.metro_distance_km;
"""

metro_time_analysis = pd.read_sql_query(query, conn)
print(metro_time_analysis)


     metro_distance_km  avg_days_on_market
0                 0.21          108.001634
1                 0.22           81.007696
2                 0.24           21.994073
3                 0.25           57.999647
4                 0.26           99.005723
..                 ...                 ...
458               7.93          102.010227
459               7.94           81.996154
460               7.95           71.666221
461               7.99           64.999727
462               8.00           70.338607

[463 rows x 2 columns]


metro distance does not strongly affect how quickly properties sell in this dataset.

What is the monthly sales trend?

In [94]:
query = """
SELECT 
    substr(date_sold, 1, 7) AS month,
    COUNT(*) AS total_sales
FROM sales
GROUP BY month
ORDER BY month;
"""

monthly_sales = pd.read_sql_query(query, conn)
print(monthly_sales)


      month  total_sales
0   2023-01            7
1   2023-02           17
2   2023-03           20
3   2023-04           36
4   2023-05           45
5   2023-06           31
6   2023-07           48
7   2023-08           50
8   2023-09           46
9   2023-10           64
10  2023-11           46
11  2023-12           42
12  2024-01           44
13  2024-02           44
14  2024-03           39
15  2024-04           29
16  2024-05           45
17  2024-06           29
18  2024-07           24
19  2024-08           13
20  2024-09            1


Sales grew through 2023, reached a peak in October 2023, and then gradually declined during 2024.

Which properties are currently unsold?

In [95]:
query = """
SELECT *
FROM listings
WHERE listing_id NOT IN (
    SELECT listing_id FROM sales
);
"""

unsold_properties = pd.read_sql_query(query, conn)
print(unsold_properties)


      listing_id         city property_type         price         sqft  \
0         L00002  Los Angeles     Apartment  1.519141e+06  4966.988193   
1         L00005      Phoenix     Townhouse  5.622970e+05  4178.997421   
2         L00009     New York     Townhouse  1.224551e+06  1522.007780   
3         L00014      Chicago     Apartment  1.377440e+06  1900.998129   
4         L00015  Los Angeles         House  1.069429e+06  3607.995839   
...          ...          ...           ...           ...          ...   
20475     L21196      Phoenix         House  7.093553e+05   898.410000   
20476     L21197      Phoenix         Condo  3.288420e+06  4195.680000   
20477     L21198  Los Angeles     Apartment  1.667441e+06  8723.370000   
20478     L21199  Los Angeles     Townhouse  1.648412e+06  8412.710000   
20479     L21200     New York     Apartment  4.991528e+06  3994.920000   

               date_listed agent_id   latitude   longitude  
0      2023-02-14 00:00:00    A0038  42.547892  -9

A total of 20,480 properties are currently unsold

Which agents have closed the most sales?

In [97]:
query = """
SELECT 
    a.agent_id,
    a.name,
    COUNT(s.listing_id) AS total_sales_closed
FROM listings l
JOIN sales s
    ON l.listing_id = s.listing_id
JOIN agents a
    ON l.agent_id = a.agent_id
GROUP BY a.agent_id, a.name
ORDER BY total_sales_closed DESC;
"""

top_agents = pd.read_sql_query(query, conn)
print(top_agents)


   agent_id         name  total_sales_closed
0     A0042  Agent A0042                  25
1     A0011  Agent A0011                  24
2     A0014  Agent A0014                  21
3     A0035  Agent A0035                  21
4     A0043  Agent A0043                  20
5     A0046  Agent A0046                  20
6     A0007  Agent A0007                  19
7     A0048  Agent A0048                  19
8     A0027  Agent A0027                  18
9     A0029  Agent A0029                  18
10    A0009  Agent A0009                  17
11    A0019  Agent A0019                  17
12    A0021  Agent A0021                  17
13    A0025  Agent A0025                  17
14    A0050  Agent A0050                  17
15    A0003  Agent A0003                  16
16    A0015  Agent A0015                  16
17    A0033  Agent A0033                  16
18    A0036  Agent A0036                  16
19    A0039  Agent A0039                  16
20    A0001  Agent A0001                  15
21    A000

Agent A0042 has closed the most sales with 25 sales.

Who are the top agents by total sales revenue?

In [98]:
query = """
SELECT 
    a.agent_id,
    a.name,
    SUM(s.sale_price) AS total_sales_revenue
FROM listings l
JOIN sales s
    ON l.listing_id = s.listing_id
JOIN agents a
    ON l.agent_id = a.agent_id
GROUP BY a.agent_id, a.name
ORDER BY total_sales_revenue DESC;
"""

top_agents_revenue = pd.read_sql_query(query, conn)
print(top_agents_revenue)


   agent_id         name  total_sales_revenue
0     A0011  Agent A0011         2.788227e+07
1     A0042  Agent A0042         2.719161e+07
2     A0043  Agent A0043         2.410242e+07
3     A0035  Agent A0035         2.272575e+07
4     A0014  Agent A0014         2.203401e+07
5     A0046  Agent A0046         2.133580e+07
6     A0048  Agent A0048         2.118631e+07
7     A0027  Agent A0027         2.109970e+07
8     A0009  Agent A0009         2.027927e+07
9     A0029  Agent A0029         1.958650e+07
10    A0039  Agent A0039         1.922301e+07
11    A0036  Agent A0036         1.920954e+07
12    A0050  Agent A0050         1.896855e+07
13    A0025  Agent A0025         1.881859e+07
14    A0017  Agent A0017         1.858385e+07
15    A0007  Agent A0007         1.841448e+07
16    A0001  Agent A0001         1.817392e+07
17    A0015  Agent A0015         1.775286e+07
18    A0019  Agent A0019         1.673061e+07
19    A0021  Agent A0021         1.654831e+07
20    A0006  Agent A0006         1

In [99]:
query = """
SELECT 
    l.agent_id,
    AVG(s.days_on_market) AS avg_days_to_close
FROM listings l
JOIN sales s
    ON l.listing_id = s.listing_id
GROUP BY l.agent_id
ORDER BY avg_days_to_close ASC;
"""

fastest_agents = pd.read_sql_query(query, conn)
print(fastest_agents)


   agent_id  avg_days_to_close
0     A0044          36.877523
1     A0013          42.428800
2     A0037          44.500154
3     A0002          45.505647
4     A0008          46.636009
5     A0017          48.566175
6     A0026          49.502797
7     A0012          50.637417
8     A0010          51.131418
9     A0043          52.351549
10    A0009          52.708242
11    A0018          54.693549
12    A0022          55.555781
13    A0028          55.599802
14    A0040          55.923827
15    A0041          56.135224
16    A0036          56.376269
17    A0023          56.502256
18    A0046          59.896999
19    A0027          61.333102
20    A0047          61.358609
21    A0030          61.546377
22    A0049          61.802167
23    A0001          61.934769
24    A0019          62.002528
25    A0029          62.003297
26    A0024          62.214727
27    A0020          62.301740
28    A0032          62.501184
29    A0033          62.749932
30    A0050          62.880843
31    A0

In [100]:
query = """
SELECT 
    a.agent_id,
    a.experience_years,
    COUNT(s.listing_id) AS total_deals_closed
FROM listings l
JOIN sales s
    ON l.listing_id = s.listing_id
JOIN agents a
    ON l.agent_id = a.agent_id
GROUP BY a.agent_id, a.experience_years
ORDER BY a.experience_years;
"""

experience_vs_deals = pd.read_sql_query(query, conn)
print(experience_vs_deals)


   agent_id  experience_years  total_deals_closed
0     A0035                 1                  21
1     A0015                 2                  16
2     A0033                 3                  16
3     A0007                 4                  19
4     A0040                 4                  13
5     A0004                 5                  11
6     A0027                 5                  18
7     A0041                 5                  15
8     A0006                 7                  15
9     A0034                 7                   7
10    A0024                 9                  14
11    A0001                10                  15
12    A0046                10                  20
13    A0050                11                  17
14    A0038                12                  12
15    A0022                13                   9
16    A0030                13                  11
17    A0042                13                  25
18    A0013                14                   7


there is no strong correlation between experience and deals closed

Do agents with higher ratings close deals faster?

In [101]:
query = """
SELECT 
    a.agent_id,
    a.rating,
    AVG(s.days_on_market) AS avg_days_to_close
FROM listings l
JOIN sales s
    ON l.listing_id = s.listing_id
JOIN agents a
    ON l.agent_id = a.agent_id
GROUP BY a.agent_id, a.rating
ORDER BY a.rating DESC;
"""

rating_vs_speed = pd.read_sql_query(query, conn)
print(rating_vs_speed)


   agent_id  rating  avg_days_to_close
0     A0020     5.0          62.301740
1     A0026     5.0          49.502797
2     A0006     4.9          81.333754
3     A0011     4.9          64.707348
4     A0013     4.8          42.428800
5     A0016     4.8          64.537198
6     A0029     4.8          62.003297
7     A0040     4.8          55.923827
8     A0027     4.6          61.333102
9     A0004     4.5          80.087771
10    A0009     4.4          52.708242
11    A0001     4.3          61.934769
12    A0005     4.3          63.901358
13    A0017     4.3          48.566175
14    A0038     4.3          70.331742
15    A0042     4.3          66.998595
16    A0045     4.2          69.078485
17    A0015     4.1          65.941069
18    A0018     4.1          54.693549
19    A0023     4.1          56.502256
20    A0024     4.1          62.214727
21    A0037     4.1          44.500154
22    A0049     4.1          61.802167
23    A0028     4.0          55.599802
24    A0047     4.0      

There is no clear relationship between agent rating and speed of closing deals.



What is the average commission earned by each agent?

In [107]:
query = """
SELECT 
    a.agent_id,
    AVG(s.sale_price * (a.commission_rate / 100.0)) AS avg_commission
FROM sales s
JOIN listings l 
    ON s.listing_id = l.listing_id
JOIN agents a 
    ON l.agent_id = a.agent_id
GROUP BY a.agent_id
ORDER BY avg_commission DESC;
"""

result = pd.read_sql_query(query, conn)
print(result)


   agent_id  avg_commission
0     A0009    33997.608030
1     A0002    32326.156470
2     A0035    32032.488514
3     A0012    31737.607165
4     A0030    31665.092319
5     A0018    31438.364525
6     A0031    30358.050323
7     A0050    28564.401636
8     A0017    28406.743728
9     A0024    27805.895161
10    A0003    27760.699311
11    A0019    27556.302083
12    A0032    27379.032259
13    A0029    26659.400085
14    A0021    26574.635809
15    A0020    25646.693076
16    A0034    24640.041816
17    A0042    24363.678966
18    A0001    24231.893310
19    A0049    23723.406508
20    A0011    23351.402817
21    A0013    23176.667989
22    A0023    23049.926473
23    A0045    22728.984698
24    A0033    22723.438330
25    A0027    22506.343425
26    A0025    22250.216479
27    A0028    22121.754765
28    A0043    21089.615770
29    A0036    20770.319389
30    A0014    20669.998000
31    A0010    20298.847301
32    A0004    19364.053905
33    A0044    18573.680787
34    A0037    18267

Which agents currently have the most active listings?

In [109]:
query = """
SELECT 
    l.agent_id,
    COUNT(*) AS active_listings
FROM listings l
LEFT JOIN sales s
    ON l.listing_id = s.listing_id
WHERE s.listing_id IS NULL
GROUP BY l.agent_id
ORDER BY active_listings DESC;
"""

result = pd.read_sql_query(query, conn)
print(result)



   agent_id  active_listings
0     A0023              446
1     A0011              439
2     A0008              438
3     A0042              435
4     A0014              432
5     A0044              430
6     A0020              426
7     A0048              425
8     A0015              425
9     A0012              425
10    A0016              424
11    A0036              422
12    A0031              422
13    A0018              422
14    A0033              418
15    A0022              417
16    A0040              416
17    A0027              415
18    A0010              415
19    A0002              414
20    A0037              413
21    A0026              412
22    A0005              412
23    A0007              411
24    A0045              410
25    A0035              410
26    A0050              409
27    A0021              409
28    A0009              409
29    A0043              408
30    A0028              407
31    A0004              406
32    A0047              404
33    A0032   

What percentage of buyers are investors vs end users?

In [111]:
query = """
SELECT 
    buyer_type,
    COUNT(*) * 100.0 / (SELECT COUNT(*) FROM buyers) AS percentage
FROM buyers
GROUP BY buyer_type;
"""

result = pd.read_sql_query(query, conn)
print(result)


  buyer_type  percentage
0   End User        50.1
1   Investor        49.9


In [113]:
query = """
SELECT 
    l.city,
    SUM(b.loan_taken) * 100.0 / COUNT(*) AS loan_uptake_rate
FROM buyers b
JOIN listings l 
    ON b.sale_id = l.listing_id
GROUP BY l.city
ORDER BY loan_uptake_rate DESC;
"""

result = pd.read_sql_query(query, conn)
print(result)


          city  loan_uptake_rate
0  Los Angeles         50.985360
1      Chicago         50.566212
2     New York         50.223156
3      Houston         49.711459
4      Phoenix         49.497111


Los Angeles has the highest loan uptake, but the difference is very small.

What is the average loan amount by buyer type?

In [114]:
query = """
SELECT 
    buyer_type,
    AVG(loan_amount) AS avg_loan_amount
FROM buyers
WHERE loan_taken = 1
GROUP BY buyer_type;
"""

result = pd.read_sql_query(query, conn)
print(result)


  buyer_type  avg_loan_amount
0   End User     5.210307e+06
1   Investor     5.211546e+06


Both groups borrow roughly the same average loan amount.

Which payment mode is most commonly used?

In [115]:
query = """
SELECT 
    payment_mode,
    COUNT(*) AS total_transactions
FROM buyers
GROUP BY payment_mode
ORDER BY total_transactions DESC;
"""

result = pd.read_sql_query(query, conn)
print(result)


    payment_mode  total_transactions
0           Cash                5088
1            UPI                5012
2         Cheque                4951
3  Bank Transfer                4949


Cash is the most commonly used payment mode, but all payment methods are used in almost equal proportions.

Do loan-backed purchases take longer to close?

In [116]:
query = """
SELECT 
    b.loan_taken,
    AVG(s.days_on_market) AS avg_days_to_close
FROM buyers b
JOIN listings l 
    ON b.sale_id = l.listing_id
JOIN sales s 
    ON l.listing_id = s.listing_id
GROUP BY b.loan_taken;
"""

result = pd.read_sql_query(query, conn)
print(result)


  loan_taken  avg_days_to_close
0          0          61.070951
1          1          62.146562


Loan-backed purchases take slightly longer to close , but the difference is minimal